### ISSUE TITLE
Non-determinism in RelGraphConv foward

### ISSUE THREAD

**Issue by ZhaofengWu** [dmlc/dgl#1471]
I noticed that if I call `RelGraphConv`'s `forward` multiple times with the same graph and the same data, there is some small variation between consecutive calls. What could be the reason? Is there a way to control that? Thanks!
```
>>> type(layer)
<class 'dgl.nn.pytorch.conv.relgraphconv.RelGraphConv'>
>>> layer.dropout
Dropout(p=0.0, inplace=False)
>>> sum(v.item() for v in layer(graph, h, graph.edata['type'], graph.edata['norm']).reshape(-1))
165742.11352539062
>>> sum(v.item() for v in layer(graph, h, graph.edata['type'], graph.edata['norm']).reshape(-1))
165742.11431884766
>>> sum(v.item() for v in layer(graph, h, graph.edata['type'], graph.edata['norm']).reshape(-1))
165742.11352539062
>>> sum(v.item() for v in layer(graph, h, graph.edata['type'], graph.edata['norm']).reshape(-1))
165742.11401367188
```

**Comment by ZhaofengWu** [comment]
This is happening on both CPU and GPU.

**Comment by yzh119** [comment]
Hi @ZhaofengWu , dgl use atomic operators(https://github.com/dmlc/dgl/blob/master/src/kernel/cuda/atomic.cuh) when you call 
 update_all with builtin message+reduce functions, and there is no way to guarantee determinism in such cases, see https://pytorch.org/docs/stable/notes/randomness.html#pytorch.

[This PR](https://github.com/dmlc/dgl/pull/1137) can address the determinism problem and we plan to pick it up again, you can expect determinism in the next release.

**Comment by ZhaofengWu** [comment]
Thanks! May I know the timeline of the next release?

**Comment by jermainewang** [comment]
@ZhaofengWu Currently we are evaluating the workload. We don't have a specific timeline yet, but please stay tune!

**Comment by yzh119** [comment]
@ZhaofengWu , the issue have been resolved in DGL 0.5.

**Comment by VickiCui** [comment]
> @ZhaofengWu , the issue have been resolved in DGL 0.5.

It seems that this issue still exists in DGL 0.5. After several steps, the output of RelGraphConv will be different with the same data and the same random seed.

**Comment by lwaekfjlk** [comment]
According to my observation, this mysterious phenomenon can be caused by the numeric error caused by simple float32 (used in pytorch)
Before calling dgl.update_all, we need to convert the related tensors into *double* type and convert them back to *float32* type after updata_all. These simple two lines perfectly solve similar problems in my code when the gradient of my model is quite small and my results are sensitive the the change of loss.

**Comment by VoVAllen** [comment]
@lwaekfjlk We've also discovered the numerical instability in RGCN model, but didn't find the root cause. Could you elaborate more on your solution? And do you have any more ideas on this problem?

**Comment by lwaekfjlk** [comment]
My numerical instability problems appear when I try to use GraphConv and SAGEConv, I am not familiar with RGCN model, but I think maybe my solution can be helpful to you.
My solution is that whatever Conv in the GNN, when doing update_all with float32 (default in Pytorch) can cause some mysterious instability. As a result, when I try to do SAGEConv with mean,  I modified the source code in  SAGEConv.py to pass double tensor into the update_all function and change it back to float32 tensor after the calculation process of update_all. I guess the reason might be some tiny operations in the update_all can be inaccurate when it comes to grad or tensors below 1e-6? I just have a simple guess and do not know the low-level true root cause either.
`graph.srcdata['h'] = graph.srcdata['h'].double()
graph.update_all(msg_fn, fn.mean('m', 'neigh'))
graph.dstdata['neigh'] = graph.dstdata['neigh'].float()`

**Comment by fmellomascarenhas** [comment]
I have a graph that is not deterministic when doing graph.update_all.

- I can confirm that using .double() before the message pass and .float() after it fixes it;
- I can also say that I was not able to recreate any graph that has the same behavior;
- My graph has this behavior when I
1. Send the graph .to('cuda');
2. Divide an int32 variable by a floating number;
3. Do the message pass

If I send to cuda after dividing, or if I transform the int32 variable into float using .float(), the message pass is deterministic.

To reproduce:
Download the graph: https://drive.google.com/file/d/1LwmoATh5e4jtK3GlFUdJHcgtmpKo1Unr/view?usp=sharing

```
import torch
import dgl.function as fn
s = torch.load('stochastic_graph.pt').to('cuda')
s.ndata['center_dist'] = s.ndata['center_dist']/5.5445

s.update_all(fn.copy_u('center_dist', 'm'), fn.sum('m', 'n'))
n = s.ndata['n'].clone()
s.update_all(fn.copy_u('center_dist', 'm'), fn.sum('m', 'n'))
n2 = s.ndata['n'].clone()

(n != n2).sum()
```

DGL version: 0.8.0post1
Pytorch version: 1.10.0
CUDA: 11.4

**Comment by VickiCui** [comment]
您好，您的邮件我已经收到，我会第一时间处理。
